R/assert.R

```diff
@@ -84,3 +93,71 @@ vec_is <- function(x, ptype = NULL, size = NULL) {
 
   TRUE
 }
+
+#' Is an object a vector?
+#'
+#' @description
+#'
+#' A vector is a collection of values like a logical vector, a factor,
+#' or a date vector. Vectors can be [sliced][vec_slice] and can be
+#' assigned as data frame columns. The opposite of a vector is a
+#' scalar. For example, a function is not a vector, and neither is a
+#' linear model. An environment is a collection of values but it
+#' cannot be a data frame column and so is not a vector.
+#'
+#' Compared to base R our definition is both more generic and
+#' constrained. In base R, anything that is an atomic vector or a list
+#' is a vector, even a model fit object ([base::is.vector()] returns
+#' `TRUE` for these objects). For the most part, functions calls
+#' implement the vector interface and can generally be treated as
+#' vectors. In vctrs, all of these objects are scalars.
+#'
+#' @section Dispatch:
+#'
+#' This function is internally generic:
+#'
+#' * Atomic vectors are always vectors.
+#' * Bare lists are vectors.
+#' * Data frames are always vectors.
+#' * Non-vector base types are always scalars.
+#'
+#' Only lists can be overriden by implementing an S3 method for
```

Comment:
The behaviour of `vec_is_vector()` can be overridden for S3 objects built on top of lists by providing a method.

Reply:
And recommend that you build on top of `vctrs_vctr` for new classes.